### PR TITLE
read_qc subworkflow

### PR DESCRIPTION
**Pull request by chrisAta**
Adding subworkflow that performs read QC using two modules:
- fastp (imported from nf-core, modified slightly to remove parameters we don't need)
- seqtk/seq (imported from nf-core)

### CODE REVIEW

**Comment by KateSakharova** [comment]
General question. If you took a module from nf-core should it be under `nf-core` folder rather than `ebi-metagenomics`?

**Comment by chrisAta** [comment]
> General question. If you took a module from nf-core should it be under `nf-core` folder rather than `ebi-metagenomics`?

Good question... I think from what @mberacochea told me before you can't have modules from more than one 'nf-core repo' for a custom modules repo, so I think the nf-core tools would break if we tried. For example trying to install nf-core/fastp:

```
nf-core modules install fastp

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 2.10.dev0 - https://nf-co.re


ERROR    You cannot install a fastp in a clone of nf-core/modules  
```

And I think some of the nf-core tools look at `org_path` in `.nf-core.yml` to know where to look for modules, and in this repo we have it set to `ebi-metagenomics`, so if we manually created an nf-core directory I don't think it would look there anyway

**Comment by KateSakharova** [comment]
@chrisAta OK, I see. 
In my vision we should definetely distinguish nf-core taken tools and our self-written. We need to mark it somehow ir create a symlink maybe? 
I'm just thinking it can be a case we bump into very BAD nf-core written module and we want to re-create it completely from our side (less likly but who knows...) what we should do?
Having nf-core subfolder can resolve it

Those are just are my thoughts for future. No need to do anything now :)

File: subworkflows/ebi-metagenomics/read_qc/main.nf [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+workflow  READ_QC {
+
+    take:
+    ch_reads // channel: [ val(meta), [ fastq ] ]
+
+    main:
+
+    ch_versions = Channel.empty()
+
+    save_trimmed_fail = true
```
Comment (KateSakharova) [inline subworkflows/ebi-metagenomics/read_qc/main.nf:14]: I would recommend to move `save_trimmed_fail` and `save_merged` to config and use as `params.save_trimmed_fail` and `params.save_merged` 
like https://github.com/nf-core/mag/blob/master/subworkflows/local/binning.nf#L69 
I took this idea from one of nf-core subwfs

**Comment by chrisAta** [comment]
> @chrisAta OK, I see. In my vision we should definetely distinguish nf-core taken tools and our self-written. We need to mark it somehow or create a symlink maybe? I'm just thinking it can be a case we bump into very BAD nf-core written module and we want to re-create it completely from our side (less likly but who knows...) what we should do? Having nf-core subfolder can resolve it
> 
> Those are just are my thoughts for future. No need to do anything now :)

Yes I completely agree - would be nice to have a more obvious and defined way for making it clear we didn't build a module from scratch 😄 Something to discuss in near future

**Review by KateSakharova — CHANGES_REQUESTED** [review]

File: modules/ebi-metagenomics/fastp/main.nf [hunk 2]
```diff
@@ -0,0 +1,102 @@
+// This fastp module is simply copied from the already-existing nf-core module (https://nf-co.re/modules/fastp)
+// This is because there are not currently any nf-core ways of adding modules from more than one nf-core repo
+// One slight change to it compared to the original is I've removed the "adapter_fasta" input as we are unlikely
+// to need it for our purposes
+
+process FASTP {
```
Comment (mberacochea) [inline modules/ebi-metagenomics/fastp/main.nf:6]: We should include the commit hash for the version of fastp we copied over -> https://github.com/nf-core/modules/commit/d497a4868ace3302016ea8ed4b395072d5e833cd

File: modules/ebi-metagenomics/seqtk/seq/main.nf [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// This seqtk/seq module is simply copied from the already-existing nf-core module (https://nf-co.re/modules/seqtk_seq/)
+// This is because there are not currently any nf-core ways of adding modules from more than one nf-core repo
+
```
Comment (mberacochea) [inline modules/ebi-metagenomics/seqtk/seq/main.nf:3]: Same here, please include in the comment the hash of the commit

File: subworkflows/ebi-metagenomics/read_qc/main.nf [hunk 4]
```diff
@@ -0,0 +1,30 @@
+
+include { FASTP      } from '../../../modules/ebi-metagenomics/fastp/main'
+include { SEQTK_SEQ     } from '../../../modules/ebi-metagenomics/seqtk/seq/main'
+
+workflow  READ_QC {
```
Comment (mberacochea) [inline subworkflows/ebi-metagenomics/read_qc/main.nf:5]: Shouldn't be "READS_QC" (plural)?

File: subworkflows/ebi-metagenomics/read_qc/main.nf [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    save_merged = true
+
+    FASTP ( ch_reads, save_trimmed_fail, save_merged )
+    ch_versions = ch_versions.mix(FASTP.out.versions.first())
+
+    SEQTK_SEQ ( FASTP.out.reads_merged )
+    ch_versions = ch_versions.mix(SEQTK_SEQ.out.versions.first())
+
+    emit:
+    reads      = FASTP.out.reads           // channel: [ val(meta), [ fastq ] ]
```
Comment (mberacochea) [inline subworkflows/ebi-metagenomics/read_qc/main.nf:24]: Please align the assignations:
```nextflow
```suggestion
    reads         = FASTP.out.reads         // channel: [ val(meta), [ fastq ] ]
    reads_merged  = FASTP.out.reads_merged  // channel: [ val(meta), [ fastq ] ]
    qc_stats      = FASTP.out.json          // channel: [ val(meta), [ json ] ]
    reads_fasta   = SEQTK_SEQ.out.fastx     // channel: [ val(meta), [ fasta ] ]
```
```

File: tests/subworkflows/ebi-metagenomics/read_qc/data/SRR21814853_1.fastq.gz
Comment (mberacochea) [inline tests/subworkflows/ebi-metagenomics/read_qc/data/SRR21814853_1.fastq.gz:1]: At some point we should consolidate the testing data, SRR21814853_{1,2}.fastq.gz is duplicated.

File: tests/subworkflows/ebi-metagenomics/read_qc/test.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    - subworkflows/read_qc
+  files:
+    - path: output/fastp/test.fastp.html
+      contains:
+        - "1.789477 M (97.392334%)"
+    - path: output/fastp/test.fastp.json
+      md5sum: 552903446c04b38bd8f2f8976119fc42
+    - path: output/fastp/test.fastp.log
+      contains:
+        - "Q30 bases: 6864820(64.3075%)"
```
Comment (mberacochea) [inline tests/subworkflows/ebi-metagenomics/read_qc/test.yml:17]: :nice:

**Review by mberacochea — CHANGES_REQUESTED** [review]
Great stuff @chrisAta .. I've added some minor comments.

File: subworkflows/ebi-metagenomics/read_qc/main.nf (hunk 1 above)
Comment (chrisAta) [reply subworkflows/ebi-metagenomics/read_qc/main.nf]: Done! Thanks for reviewing 😄

**Comment by chrisAta** [comment]
> Great stuff @chrisAta .. I've added some minor comments.

All done! Thanks for the review 😄

File: tests/subworkflows/ebi-metagenomics/reads_qc/nextflow.config [hunk 7]
```diff
@@ -0,0 +1,14 @@
+process {
+
+    publishDir = { "${params.outdir}/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" }
+
+    withName: SEQTK_SEQ {
```
Comment (mberacochea) [inline tests/subworkflows/ebi-metagenomics/reads_qc/nextflow.config:5]: This one should be also included in a nextflow.config file in the subworkflow root folder as https://github.com/nf-core/modules/blob/master/subworkflows/nf-core/bam_docounts_contamination_angsd/nextflow.config

**Review by mberacochea — CHANGES_REQUESTED** [review]
One last bit.. the nextflow.config file for the subworkflow.

File: tests/subworkflows/ebi-metagenomics/reads_qc/nextflow.config (hunk 7 above)
Comment (mberacochea) [reply tests/subworkflows/ebi-metagenomics/reads_qc/nextflow.config:5]: I don't think so.. it's just for the key "-a" parameter

**Review by mberacochea — APPROVED** [review]
Great stuff :shipit:

**Review by KateSakharova — APPROVED** [review]

**Comment by KateSakharova** [comment]
Thank you Chris!

**Pull request merged**